### PR TITLE
Feature/issue 481 append array vectors of vectors

### PR DESCRIPTION
**Pull request by bbbales2**
#### Submission Checklist

- [only the append_array_test ones] Run unit tests: `./runTests.py test/unit`
- [x] Run cpplint: `make cpplint`
- [x] Declare copyright holder and open-source license: see below

#### Summary:

This replaces this pull req: https://github.com/stan-dev/math/pull/575 (sorry bout that -- lost my commit permissions to Stan repos last week -- Sean n' Daniel said to just make a pull req. from my own fork).

Two questions @bob-carpenter:
1. I expanded the new tests to test all the different ways to call append_array (up to two std::vectors deep). To avoid duplication of the test code, I put all the tests in /mix/ instead of splitting them up into individual sections. That okay?

2. These new tests only test that values are being passed around correctly (either the actual value of a double, .val() of var, .val().val() of fvar<var>, etc.). I'm not testing the autodiff. I'm not testing values of derivatives in fvars. That said, I think that's about all the append_array function should really do. I can delete all the old append_array tests that test the gradients and such and I don't think we're really losing anything (here are your previous thoughts on this: https://github.com/stan-dev/math/pull/531#issuecomment-298264800)

Response to req. changes:
1. I left the return-by-reference builders in place in the tests. I found it easier to work with templates like this.

#### How to Verify:

./runTests.py test/unit/math/prim/mat/fun/append_array_test.cpp
./runTests.py test/unit/math/fwd/mat/fun/append_array_test.cpp
./runTests.py test/unit/math/rev/mat/fun/append_array_test.cpp
./runTests.py test/unit/math/mix/mat/fun/append_array_test.cpp

#### Documentation:

In the Doxygen

#### Reviewer Suggestions: 

@bob-carpenter 

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): University of California, Santa Barbara

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

File: stan/math/prim/mat/fun/append_array.hpp [hunk 1]
```diff
@@ -1,80 +1,141 @@
...
 #include <stan/math/prim/mat/fun/Eigen.hpp>
-#include <stan/math/prim/scal/meta/return_type.hpp>
+#include <stan/math/prim/mat/meta/append_return_type.hpp>
+#include <stan/math/prim/scal/err/check_consistent_sizes.hpp>
 #include <vector>
 
 namespace stan {
   namespace math {
+    /**
+     * This function copies the value at x to y, casting from the type of T1 to
```
Comment (mitzimorris) [inline stan/math/prim/mat/fun/append_array.hpp]: comment needs context - explain when append_array reduced to copy x into y.

File: stan/math/prim/mat/fun/append_array.hpp [hunk 2]
```diff
@@ -1,80 +1,141 @@
...
+     */
+    template<typename T1, typename T2>
+    inline void append_array_copy(const T1& x, T2& y) {
+      y = T2(x);
+    }
 
     /**
-     * Return the concatenation of two specified integer vectors in the
-     *   order of the arguments.
+     * This copies and casts the elements of x to y (by delegating to the Eigen
```
Comment (mitzimorris) [inline stan/math/prim/mat/fun/append_array.hpp]: same as first comment

File: stan/math/prim/mat/fun/append_array.hpp [hunk 3]
```diff
@@ -1,80 +1,141 @@
...
+    template<typename T1, typename T2, int R, int C>
+    inline void append_array_copy(const Eigen::Matrix<T1, R, C>& x,
+                     Eigen::Matrix<T2, R, C>& y) {
+      y = x.template cast<T2>();
     }
 
     /**
-     * Return the concatenation of two vectors containing Matrices
-     *   of the same type in the order of the arguments.
+     * This function copies and casts the elements of std::vector x to std::vector y.
```
Comment (mitzimorris) [inline stan/math/prim/mat/fun/append_array.hpp]: same as first comment

File: stan/math/prim/mat/fun/append_array.hpp (hunk 1 above)
Comment (bob-carpenter) [inline stan/math/prim/mat/fun/append_array.hpp]: I'm not going to double-review this, but I will pack all my notes into a review so as not to spam everyone on our lists.  

I think this function just needs to be renamed to something like `convert_copy` --- I'm guessing @mitzimorris was thinking this was actually appending arrays.

File: stan/math/prim/mat/fun/append_array.hpp [hunk 4]
```diff
@@ -1,80 +1,141 @@
...
-    inline std::vector<int>
-    append_array(const std::vector<int>& x, const std::vector<int>& y) {
-      std::vector<int> z;
-      z.reserve(x.size() + y.size());
-      z.insert(z.end(), x.begin(), x.end());
-      z.insert(z.end(), y.begin(), y.end());
-      return z;
+    template<typename T1, typename T2, int R, int C>
+    inline void append_array_copy(const Eigen::Matrix<T1, R, C>& x,
+                     Eigen::Matrix<T2, R, C>& y) {
```
Comment (bob-carpenter) [inline stan/math/prim/mat/fun/append_array.hpp]: Second arguments should be lined up under the first unless they won't fit on a line.

File: stan/math/prim/mat/fun/append_array.hpp [hunk 5]
```diff
@@ -1,80 +1,141 @@
...
-      if (!x.empty() && !y.empty()) {
-        check_matching_dims("append_array",
-                            "x[1]",
-                            x.front(),
-                            "y[1]",
-                            y.front());
+    template<typename T1, typename T2>
+    inline void append_array_copy(const std::vector<T1> &x,
+                                  std::vector<T2> &y) {
+      y.resize(x.size());
```
Comment (bob-carpenter) [inline stan/math/prim/mat/fun/append_array.hpp]: What are the invariants here?  Is it possible to call this without `x` and `y` being the same size given your checks above?  Those just checked first elements, so are you anticipating ragged uses here or something (in which case those first element checks are wrong).

File: stan/math/prim/mat/fun/append_array.hpp [hunk 6]
```diff
@@ -1,80 +1,141 @@
...
 
-      std::vector<Eigen::Matrix<typename return_type<T1, T2>::type, R, C> > z;
-      z.reserve(x.size() + y.size());
-      for (size_t i = 0; i < x.size(); i++)
-        z.push_back(x[i].template cast<typename return_type<T1, T2>::type>());
-      for (size_t i = 0; i < y.size(); i++)
-        z.push_back(y[i].template cast<typename return_type<T1, T2>::type>());
-      return z;
+    /**
+     * This function checks the sizes of both arguments are consistent.
```
Comment (bob-carpenter) [inline stan/math/prim/mat/fun/append_array.hpp]: What does it do if they're not?

File: stan/math/prim/mat/fun/append_array.hpp [hunk 7]
```diff
@@ -1,80 +1,141 @@
...
+     *
+     * The basic version assumes variables are of compatible sizes. There are
+     * specializations for Eigen::Matrices and std::vectors (where this
```
Comment (bob-carpenter) [inline stan/math/prim/mat/fun/append_array.hpp]: There is no Eigen::Matrices, it's Eigen::Matrix.  I also didn't understand this comment.  If the variables aren't `Eigen::Matrix or `std::vector` types, then they're scalars and have to be the same size.   Did you just mean "scalar" instead of "compatible sizes"?

What happens if you pass in one scalar and one vector?  Or is that never going to happen because of the outside calling structure?  If so, you might want to say that there's an underlying assumption the two arguments are of the same shape, if not the same scalar type.

There's no notion of "basic version".  This can be a base template function (which can get specialized) or just a standalone overload.

File: stan/math/prim/mat/fun/append_array.hpp [hunk 8]
```diff
@@ -1,80 +1,141 @@
...
+     * @tparam T1 Scalar type of x
+     * @tparam T2 Scalar type of y
+     * @tparam R Eigen RowsAtCompileTime of both matrices
+     * @tparam C Eigen ColsAtCompileTime of both matrices
+     * @param x First matrix
+     * @param y Second matrix
+     */
+    template<typename T1, typename T2, int R, int C>
+    inline void append_array_check_size(const Eigen::Matrix<T1, R, C>& x,
+                     const Eigen::Matrix<T2, R, C>& y) {
```
Comment (bob-carpenter) [inline stan/math/prim/mat/fun/append_array.hpp]: Line up second args.

File: stan/math/prim/mat/fun/append_array.hpp [hunk 9]
```diff
@@ -1,80 +1,141 @@
...
+      check_matching_dims("append_array", "x", x, "y", y);
```
Comment (bob-carpenter) [inline stan/math/prim/mat/fun/append_array.hpp]: The `append_array` thing here isn't the name of the function.  Are you sure this only gets called from `append_array`?  If so, add a warning to the doc about the message saying it should only be called from `append_array`.

File: stan/math/prim/mat/fun/append_array.hpp [hunk 10]
```diff
@@ -1,80 +1,141 @@
...
+     * @tparam T1 Element type of x
+     * @tparam T2 Element type of y
+     * @param x First std::vector
+     * @param y Second std::vector
+     */
+    template<typename T1, typename T2>
+    inline void append_array_check_size(const std::vector<T1> &x,
+                                        const std::vector<T2> &y) {
+      check_consistent_sizes("append_array", "size of x", x, "size of y", y);
+      if (!x.empty() && !y.empty()) {
```
Comment (bob-carpenter) [inline stan/math/prim/mat/fun/append_array.hpp]: The `!y.empty()` check is redundant given the consistency check.  

It'd be better to make sure `append_array_check_size` worked for empty containers.  This does *not* save time to put in redundant checks like this.  Branching is expensive.

File: stan/math/prim/mat/fun/append_array.hpp [hunk 11]
```diff
@@ -1,80 +1,141 @@
...
-    inline typename std::vector<typename return_type<T1, T2>::type>
-    append_array(const std::vector<T1>& x, const std::vector<T2>& y) {
-      std::vector<typename return_type<T1, T2>::type> z;
-      z.reserve(x.size() + y.size());
-      z.insert(z.end(), x.begin(), x.end());
-      z.insert(z.end(), y.begin(), y.end());
+    inline typename append_return_type<T1, T2>::type
+    append_array(const T1& x, const T2& y) {
+      typename append_return_type<T1, T2>::type z;
+      if (!x.empty() && !y.empty()) {
```
Comment (bob-carpenter) [inline stan/math/prim/mat/fun/append_array.hpp]: same comment as above.

File: stan/math/prim/mat/fun/append_array.hpp [hunk 12]
```diff
@@ -1,80 +1,141 @@
...
+        append_array_check_size(x[0], y[0]);
+      }
+      z.resize(x.size() + y.size());
+      for (size_t i = 0; i < x.size(); i++)
+        append_array_copy(x[i], z[i]);
```
Comment (bob-carpenter) [inline stan/math/prim/mat/fun/append_array.hpp]: With this structure it's clear that you need an `append_array_copy` specializion for when the two arguments are the same type---in that case, you do *not* need to do this all recursively, and it'll be *much* faster.

File: stan/math/prim/mat/meta/append_return_type.hpp [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+  namespace math {
+    /**
+     * This template metaprogram defines the return type for appending
+     * containers of the specified template parameter types.
+     *
+     * This base implementation assumes the template parameters are not
+     * container types and calculates their return type using
+     * <code>return_type</code>.
+     *
+     * Specializations define behavior for other basic types and for containers.
```
Comment (bob-carpenter) [inline stan/math/prim/mat/meta/append_return_type.hpp]: This line (18) is redundant, but the rest of the doc's great!

File: stan/math/prim/mat/meta/append_return_type.hpp [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+     * @tparam T1 First type to be promoted
+     * @tparam T2 Second type to be promoted
+     */
+    template <typename T1, typename T2>
+    struct append_return_type {
+      typedef typename return_type<T1, T2>::type type;
+    };
+
+    /**
+     * If both types are ints, the type member is an int (this is different
```
Comment (bob-carpenter) [inline stan/math/prim/mat/meta/append_return_type.hpp]: You should still stay what this does.  

You don't need to doc differences with other functions---there's just no end to that.  You can doc the previous function to indicate it'll deal with all the scalar types other than the `(int, int)` pair.

File: stan/math/prim/mat/meta/append_return_type.hpp [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+     * @tparam T1 Scalar type of first matrix type
+     * @tparam T2 Scalar type of first matrix type
+     * @tparam R Eigen RowsAtCompileTime of both matrices
+     * @tparam C Eigen ColsAtCompileTime of both matrices
+     */
+    template <typename T1, typename T2, int R, int C>
+    struct append_return_type<Eigen::Matrix<T1, R, C>,
+                              Eigen::Matrix<T2, R, C> > {
+      typedef typename
+      Eigen::Matrix<typename append_return_type<T1, T2>::type, R, C> type;
```
Comment (bob-carpenter) [inline stan/math/prim/mat/meta/append_return_type.hpp]: Arguably what you have is more general, but from Stan's perspective, this can just be a regular `return_type` function call because we only allow `double` and autodiff scalars in matrices (not integers).

File: stan/math/prim/mat/meta/append_return_type.hpp [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+      typedef int type;
+    };
+
+    /**
+     * If both types are Eigen::Matrices with the same Row/Column specification,
+     * then the type member is another Eigen::Matrix with the same Row/Column
+     * specification and a scalar type promoted using <code>return_type</code>.
+     * Part of return type promotion logic for append_array
+     *
+     * @tparam T1 Scalar type of first matrix type
```
Comment (bob-carpenter) [inline stan/math/prim/mat/meta/append_return_type.hpp]: I'd just say `scalar type of first argument matrix`  --- I'm not sure what you meant by "first matrix type".

File: test/unit/math/fwd/mat/fun/append_array_test.cpp [hunk 17]
```diff
@@ -3,10 +3,11 @@
 #include <gtest/gtest.h>
 
 using namespace Eigen;
```
Comment (bob-carpenter) [inline test/unit/math/fwd/mat/fun/append_array_test.cpp]: This isn't about this line number, but the tests in general.  Overally, they look OK, but you want to pull repeated functionality out into functions as much as possible.  We haven't done a good job of that in tests in the past.

I don't see any `*/arr/` tests for `std::vector` or the scalar tests.  You need to make sure that int and double casting works right as well as through `std::vector<int>` and `std::vector<double>`.

**Comment by bbbales2** [comment]
Howdy Mitzi!

I made a commit to address the comments from you and Bob.

I also removed a large number of the append_array tests. There are so many different types and combinations of types that needed testing here is was kinda difficult. I made a generic test function in test/unit/math/mix/mat/fun/append_array_test.hpp that should:

1. Randomly generate two ```std::vector```s (and ```std::vector<std::vector<>>```s) of things
2. Call append_array on them
3. Check to make sure things ended up in the right spot

I removed all the autodiff checks that were in append_array. There were really an obscene number of checks for something that boils down to a copy operator. Is this okay? Also, is implementing it like this okay? I made the header in mix cause it has ```fvar<var>``` stuff in it, but I include it from prim. Also I left some usings and typedefs in the header that normally wouldn't be okay in a header file (but I think should be okay for test codes)?

Here's my responses to the last review (If I didn't respond to it, I agreed with it and made the changes):

> Mitzi's comments
> stan/math/prim/mat/fun/append_array.hpp, line 30
> comment needs context - explain when append_array reduced to copy x into y.

Yeah, it's what Bob said here. append_array_copy is just a helper function for append_array (defined at the bottom of the file).

I could change its name, but I'm hesitant to rename it something like cast_and_copy and move it out of the file because I'm not sure how it generalizes or whatever. If it becomes an official Stan Math function, then other things might depend on it and so it'd need to have consistent behavior and so forth. I'm hesitant to do that because this is just a helper function for append_array and I'd like it to stay that way.

> Bob's comments
> stan/math/prim/mat/fun/append_array.hpp, line 56
> What are the invariants here? Is it possible to call this without x and y being the same size given your checks above? Those just checked first elements, so are you anticipating ragged uses here or something (in which case those first element checks are wrong).

These are different x s and y s here. At the append_array level, this is copying elements of x and y to elements of z.

> stan/math/prim/mat/fun/append_array.hpp, line 122
> The !y.empty() check is redundant given the consistency check.
> It'd be better to make sure append_array_check_size worked for empty containers. This does not save time to put in redundant checks like this. Branching is expensive.

I made this change, but I don't super duper agree with it. check_size is lower complexity than the copies and doing this verbosely makes it clearer what's happening. I'll keep it in mind though.

> stan/math/prim/mat/fun/append_array.hpp, line 129
> same comment as above.

This doesn't apply here. x and y are the arrays to be appended together. They can have different number of elements and we only need to compare the sizes if they are not length zero.

Thanks!

**Comment by bbbales2** [comment]
By the way, the code changes for the response were pretty minor, but the test changes were pretty big. It's really easy to add another commit that rolls back the test changes if they aren't appropriate.

**Comment by bob-carpenter** [comment]
> I could change its name, but I'm hesitant to rename it something like cast_and_copy and move it out of the file because I'm not sure how it generalizes or whatever. If it becomes an official Stan Math function, then other things might depend on it and so it'd need to have consistent behavior and so forth. I'm hesitant to do that because this is just a helper function for append_array and I'd like it to stay that way.

We used to do this all the time ("just a helper function"),  which is what led to so many duplicated functions strewn throughout the code.  We made a conscious decision to put a stop to this and require either new global metaprograms or use of the existing ones. Chances are anything you need to do to append arrays will be something that's needed elsehwere as a metaprogram.

Smaller, more modular code is much much easier to maintain (doc, test, upgrade to new types, bug fix) and typically easier to understand as there are fewer functions to learn.

Yes, it's more work to pull out a function, but in the end, we decided it's worth it.  The thing is, I think there may already be a cast-and-copy like function floating around you'll find.

To sum up, check out

stan/math/prim/mat/fun/assign.hpp

You may find it does everyting you need.  We need it for upcasting int to double and these primitives to autodiff types to get around the fact that neither standard library types or Eigen vectors are covariant:

https://en.wikipedia.org/wiki/Covariance_and_contravariance_(computer_science)

> Bob's comments
> stan/math/prim/mat/fun/append_array.hpp, line 56
> What are the invariants here? Is it possible to call this without x and y being the same size given your checks above? Those just checked first elements, so are you anticipating ragged uses here or something (in which case those first element checks are wrong).
> 
> These are different x s and y s here. At the append_array level, this is copying elements of x and y to elements of z.
> 
> stan/math/prim/mat/fun/append_array.hpp, line 122
> The !y.empty() check is redundant given the consistency check.

This is minor, of course.

> It'd be better to make sure append_array_check_size worked for empty containers. This does not save time to put in redundant checks like this. Branching is expensive.
> 
> I made this change, but I don't super duper agree with it. check_size is lower complexity than the copies and doing this verbosely makes it clearer what's happening.

I disagree.  I think making the boundary conditions general makes things easier to understand.  I'm not so worried about understandability of the code---that's trivial either way in a simple case like this---more about the external interface simplicity.

> I'll keep it in mind though.

You may be surprised about the high potential cost of branches.  Simple branches like this we thought would save time in the end wound up killing the compilers ability to inline and ensure code locality.  Down at the memory management level in autodiff, you can really see the cost;  even at the level of implementing the var/vari structures you can see it.  Redundant checks at one point were literally halving the speed of autodiff.

Bottom line is that when you have to make an unpredicted branch, it's super duper expensive.  Even when you don't, it can seriously derail compiler optimizations.

If you're interested in diving deep, I'd suggest Agner Fogg's C++ optimization guides.

**Comment by bbbales2** [comment]
> stan/math/prim/mat/fun/assign.hpp

Niiice, yeah, this looks great. And more complete than what I wrote. I'll try to switch to this.

> We used to do this all the time ("just a helper function")

Just trying to be lazy :D.

How does this translate to the tests? Cause some of the code I wrote for the tests seems like it should be generalized. How to organize test headers? If they go under */mix/* then it seems weird importing them in */prim/*.

> I think making the boundary conditions general makes things easier to understand.

Whoops, yeah, I agree with this. Didn't say it quite right. I was disagreeing with the specific change.

> Simple branches like this we thought would save time in the end wound up killing the compilers ability to inline and ensure code locality.

That's pretty interesting. I was just assuming the compiler could turn !y.empty() into a couple logicals and then there'd only be one branch anyway, but that would depend on the compiler having to make sense of the C++, which I guess is a pretty brittle assumption. And it's still a couple unnecessary logicals anyway haha.

I'm definitely guilty of architecture worship. Undergrad was EE. Way back when I did a bunch of microcontroller stuff. At one point I had an FFT written in assembly with some 16 bit fixed point math that I was pretty proud of (as you can imagine that was some high precision stuff). I'm very thankful that now that I write stuff for x86 I'm not always looking up instruction documentation to make sure they're fast/doing what I want haha. The fact that there's a branch predictor to begin with always seems like some sort of miracle.

**Comment by bob-carpenter** [comment]
The dependencies in the code should only go one way

* mix -> {fwd, rev } -> prim

* mat -> arr -> scal

Right---that !y.empty() won't create another branch.  It's just a redundancy (that may also add clarity).  I just meant that branching to detect boundary conditions is often worse than just executing a loop.  Even if that loop requires an object to be constructed on the stack.  Once you need objects like std::vector or Eigen::Matrix that allocate memory on the heap, construction gets a lot more expensive.

It's OK to write tests up at the mix/mat level that test everything below it. But don't go the other way and import mix from the prim tests.

**Comment by bbbales2** [comment]
> It's OK to write tests up at the mix/mat level that test everything below it. But don't go the other way and import mix from the prim tests.

Cool beans, I've got some more fixing up to do then. Ignore this pull for now.

**Comment by bbbales2** [comment]
If the tests pass, this is ready to check again. I changed things to use the internal resize/dims as much as possible (and get rid of the helper functions I wrote). I had to make another commit to fix some minor merge conflicts between the main math develop and my own branch.

I just did: ```git merge --no-ff stan/develop``` (where stan is my remote for the Stan repos) and fixed the conflicts.

Hopefully that's the correct thing!

**Comment by bob-carpenter** [comment]
Looks like the unit tests are failing because there are too many compiler warnings.

Check the signs on comparisons and use g++ if you don't see enough warnings with clang++.

File: stan/math/prim/mat/fun/append_array.hpp [hunk 18]
```diff
@@ -1,77 +1,76 @@
...
-    inline std::vector<Eigen::Matrix<typename return_type<T1, T2>::type, R, C> >
-    append_array(const std::vector<Eigen::Matrix<T1, R, C> >& x,
-                 const std::vector<Eigen::Matrix<T2, R, C> >& y) {
-      if (!x.empty() && !y.empty()) {
-        check_matching_dims("append_array",
-                            "x[1]",
-                            x.front(),
-                            "y[1]",
-                            y.front());
+    template <typename T1, typename T2 >
```
Comment (bbbales2) [inline stan/math/prim/mat/fun/append_array.hpp]: This is the complicated version of append_array. If the scalar types in T1 and T2 do not match, then I have to do some manual casting and copying to get things appended together.

File: stan/math/prim/mat/fun/append_array.hpp [hunk 19]
```diff
@@ -1,77 +1,76 @@
...
+     * @tparam T1 Type of vectors
      * @param x First vector
      * @param y Second vector
      * @return A vector of x and y concatenated together (in that order)
      */
-    template <typename T1, typename T2>
-    inline typename std::vector<typename return_type<T1, T2>::type>
-    append_array(const std::vector<T1>& x, const std::vector<T2>& y) {
-      std::vector<typename return_type<T1, T2>::type> z;
+    template <typename T1>
```
Comment (bbbales2) [inline stan/math/prim/mat/fun/append_array.hpp]: This is the simple version of append_array where C++ can handle all the copying and doesn't need to cast things.

File: stan/math/prim/mat/fun/resize.hpp [hunk 20]
```diff
@@ -6,49 +6,32 @@
...
-        x.resize(dims[pos]);
-      }
-
-      template <typename T>
-      void resize(Eigen::Matrix<T, 1, Eigen::Dynamic>& x,
-                  const std::vector<size_t>& dims,
-                  size_t pos) {
-        x.resize(dims[pos]);
+      template <typename T, int R, int C>
+      void resize(Eigen::Matrix<T, R, C>& x,
```
Comment (bbbales2) [inline stan/math/prim/mat/fun/resize.hpp]: It didn't look like resize was being used, so I took the liberty of adopting it and modifying it to suite my needs.

Need 1 was changing it so that the resize semantics match the dim semantics. Previous, to resize a RowVector or a Vector to something like length N, you could just pass in {N}. Now to be consistent with dims, you need to pass in {1, N} for RowVector or {N, 1} for Vector (where {} are std::vectors)

Also I switched the size_ts to ints. This is also to match the dims function.

File: test/unit/math/fwd/mat/fun/resize_test.cpp [hunk 21]
```diff
@@ -7,12 +7,12 @@ using stan::math::fvar;
 
 TEST(AgradFwdMatrixResize, fvar_double) {
   fvar<double> x = 5;
-  std::vector<size_t> dims;
+  std::vector<int> dims;
```
Comment (bbbales2) [inline test/unit/math/fwd/mat/fun/resize_test.cpp]: In these tests I'm just changing size_t to int and changing how Vectors and RowVectors are resized.

File: test/unit/math/mix/mat/fun/append_array_test.cpp [hunk 22]
```diff
@@ -1,396 +1,392 @@
 #include <stan/math/mix/mat.hpp>
-#include <stan/math/prim/mat/fun/Eigen.hpp>
 #include <gtest/gtest.h>
 
+using stan::math::var;
```
Comment (bbbales2) [inline test/unit/math/mix/mat/fun/append_array_test.cpp]: This file is pretty involved. The entry points are:

```
TEST(MathFunctions, append_array_prim)
TEST(MathFunctions, append_array_rev)
TEST(MathFunctions, append_array_fwd)
TEST(MathFunctions, append_array_mix)
```

Basically they all call ```check``` with various combinations of argument types and ```check``` calls:

```checkv``` which checks appending std::vectors of those types works as expected
and
```checkvv``` which checks appending std::vectors of std::vectors of those types works as expected

There are no tests of the autodiff here. We're just testing that all the casting between various containers of different autodiff types work.

The lingo for the typedefs is:

v -> var
f -> fvar
d -> double

So ```fd``` is ```fvar<double>``` and ```ffv``` is ```fvar<fvar<double> >```

V is VectorXd, R is RowVectorXd, M is MatrixXd.

Vfd is a Vector of ```fvar<double>```.

Mffv is a Matrix of ```fvar<fvar<var> >```

File: test/unit/math/mix/mat/fun/resize_test.cpp [hunk 23]
```diff
@@ -8,12 +8,12 @@ using stan::math::var;
 
 TEST(AgradMixMatrixResize, fvar_var) {
   fvar<var> x = 5;
-  std::vector<size_t> dims;
+  std::vector<int> dims;
```
Comment (bbbales2) [inline test/unit/math/mix/mat/fun/resize_test.cpp]: Again, I'm just changing the resize tests to work with ints and use 2 dimensional sizes for Vector and RowVector.

File: test/unit/math/prim/mat/fun/append_array_test.cpp [hunk 24]
```diff
@@ -1,151 +1,69 @@
 #include <stan/math/prim/mat.hpp>
 #include <gtest/gtest.h>
 
-using namespace Eigen;
-
-TEST(MathFunctions, append_array) {
-  std::vector<double> x(3), y(2), z, result;
+TEST(MathFunctions, append_array_nan) {
```
Comment (bbbales2) [inline test/unit/math/prim/mat/fun/append_array_test.cpp]: These are just some simpler, way easier to digest tests of append_array.

File: test/unit/math/prim/mat/fun/resize_test.cpp [hunk 25]
```diff
@@ -10,12 +10,12 @@ using stan::math::row_vector_d;
 
 TEST(MathMatrix, resize_double) {
```
Comment (bbbales2) [inline test/unit/math/prim/mat/fun/resize_test.cpp]: Again, just changing resize to work with ints and match dims with the VectorXd and RowVectorXd size semantics.

File: stan/math/prim/mat/fun/append_array.hpp (hunk 18 above)
Comment (bob-carpenter) [inline stan/math/prim/mat/fun/append_array.hpp]: I don't think there literally needs to be manual casting because the autodiff var constructors are implicit and you can assign integers to real valued variables.

File: stan/math/prim/mat/fun/append_array.hpp (hunk 19 above)
Comment (bob-carpenter) [inline stan/math/prim/mat/fun/append_array.hpp]: I agree with another comment somewhere that a lot of this should just go in the doc itself.

**Comment by bob-carpenter** [comment]
Is this ready to be reviewed again @bbbales2?  If things get reviewed and require changes, it helps to let the reviewer know that it's ready to review again.  Otherwise, it's hard to tell.

**Comment by bbbales2** [comment]
Yup, my bad. It's ready for reviewin'!
